server/src/internalClusterTest/java/org/opensearch/cluster/coordination/NodeJoinLeftIT.java

```diff
@@ -0,0 +1,374 @@
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            throw new NodeHealthCheckFailureException("fake followerchecker failure simulated by test to repro race condition");
+        });
+        MockTransportService cmTransportService = (MockTransportService) internalCluster().getInstance(
+            TransportService.class,
+            clusterManager
+        );
+        MockTransportService redTransportService = (MockTransportService) internalCluster().getInstance(
+            TransportService.class,
+            redNodeName
+        );
+        redTransportService.addRequestHandlingBehavior(FOLLOWER_CHECK_ACTION_NAME, simulatedFailureBehaviour);
+
+        // Loop runs 10 times to ensure race condition gets reproduced
+        for (int i = 0; i < 10; i++) {
+            // Fail followerchecker by force to trigger node disconnect and node left
+            logger.info("--> simulating followerchecker failure to trigger node-left");
+            succeedFollowerChecker.set(false);
+            Thread.sleep(1000);
+
+            // Trigger a node disconnect while node-left task is still processing
+            logger.info(
+                "--> triggering a simulated disconnect on red node, after the follower checker failed to see how node-left task deals with this"
+            );
+            cmTransportService.disconnectFromNode(redTransportService.getLocalDiscoNode());
+
+            ClusterHealthResponse response1 = client().admin().cluster().prepareHealth().setWaitForNodes("2").get();
+            assertThat(response1.isTimedOut(), is(false));
+
+            // once we know a node has left, we can re-enable followerchecker to work normally and validate node rejoins
+            logger.info("--> re-enabling normal followerchecker and validating cluster is stable");
+            succeedFollowerChecker.set(true);
+            ClusterHealthResponse response2 = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+            assertThat(response2.isTimedOut(), is(false));
+
+            Thread.sleep(1000);
+            // checking again to validate stability and ensure node did not leave
+            ClusterHealthResponse response3 = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+            assertThat(response3.isTimedOut(), is(false));
+        }
+
+        succeedFollowerChecker.set(true);
+        ClusterHealthResponse response = client().admin().cluster().prepareHealth().setWaitForNodes("3").get();
+        assertThat(response.isTimedOut(), is(false));
+
+        // assert that the right exception message showed up in logs
+        assertTrue(
+            "Expected IllegalStateException was not logged",
+            testAppender.containsExceptionMessage("IllegalStateException[cannot make a new connection as disconnect to node")
+        );
+    }
+
+    public void testRestartDataNode() throws Exception {
```

Comment:
If the intent is to create node-left in successions, it might be good to see if we can cause n/w disruption rather than restart a node which is slow. The scenario could be something like this
- spin up a 10 node cluster
- cause a network disruption for 5 nodes (randomly)
- verify if the follower check removes
- restore the network (as the nodes are getting removed)
- wait until the cluster has 10 nodes

Reply:
The intent behind this test is mainly to just do basic sanity of data node restart and validate that nodes are able to join back in a normal case without any slowness